lib/db/mysql.js

```diff
@@ -413,92 +407,148 @@ module.exports = function (log, error) {
-      const arr = result[name] ? result[name].split(separator) : []
-      result[name] = transform ? transform(arr) : arr
-    }
-  }
-
-  function unwrapCapabilities(result, fieldName) {
-    const field = {
-      name: fieldName,
-      transform(capabilityIds) {
-        return capabilityIds.reduce((acc, c) => {
-          const capabilityStr = dbUtil.mapDeviceCapability(parseInt(c))
-          if (capabilityStr != null) {
-            acc.push(capabilityStr)
+  // A helper function for aggregating name:value pairs into a JSON object.
+  // It takes a list of rows like this:
+  //
+  //   +-------+-------+-------+--------+
+  //   | id    | other | name  | value  |
+  //   +-------+-------+-------+--------+
+  //   | one   | foo   | name1 | value1 |
+  //   | two   | bar   | name1 | value1 |
+  //   | two   | bar   | name2 | value2 |
+  //   | three | baz   | NULL  | NULL   |
+  //   +-------+-------+-------+--------+
+  //
+  // And aggregates the name/data pairs into a JSON object like this:
+  //
+  //  [
+  //    { id: 'one', other: 'foo', agg: { 'name1': 'value1' }},
+  //    { id: 'two', other: 'bar', agg: { 'name1': 'value1', 'name2': 'value1' }},
+  //    { id: 'three', other: 'baz', agg: {}},
+  //  ]
+  //
+  // Note that correct operation depends on the rows being ordered by id column,
+  // so they can be aggregated in a single pass.
+  //
+  // The column names can be customized, in order to work with several different
+  // queries.
+
+  function aggregateNameValuePairs(rows, idColumn, nameColumn, valueColumn, aggColumn) {
+    return rows.reduce((items, row) => {
+      let curItem = items[items.length - 1]
+      // Start a new aggregation item if:
+      //   * we're at the start of the list, or
+      //   * the upcoming row has a null idColumn, or
+      //   * the upcoming row has a different idColumn.
+      if (! curItem || ! row[idColumn] || ! curItem[idColumn] || ! row[idColumn].equals(curItem[idColumn])) {
+        curItem = {}
+        Object.keys(row).forEach(column => {
+          if (column !== nameColumn && column !== valueColumn) {
+            curItem[column] = row[column]
           }
-          return acc
-        }, [])
+        })
+        // When listing session tokens, we need to distinguish between
+        // a session with a deviceId but no commands (availableCommands = {})
+        // and a session with no deviceId at all (availableCommands = null).
+        curItem[aggColumn] = row[idColumn] ? {} : null
+        items.push(curItem)
       }
-    }
-    const toUnwrap = Array.isArray(result) ? result : [result]
-    toUnwrap.forEach(r => unwrapArrays(r, [field]))
-    return result
+      if (row[nameColumn]) {
+        curItem[aggColumn][row[nameColumn]] = row[valueColumn]
+      }
+      return items
+    }, [])
   }
 
-  // Select : devices d, deviceCapabilities dc, sessionTokens s, accounts a
+  // Select : devices d, sessionTokens s, deviceAvailableCommands sc, deviceCommandIdentifiers ci
```

Comment:
Just out of curiosity, what does the "s" stand for in `sc`? (the stored procedure still has it as `dc` I notice, not that it matters)

Reply:
Nice catch; these used to be "supported commands" rather than "available commands" so I guess it's leftover from that naming convention.